subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java

```diff
@@ -81,54 +82,36 @@ public static void discardListener() {
         setListener(NO_OP);
     }
 
+    private static final Map<String, CallInterceptor> GROOVY_CALLSITE_INTERCEPTORS = new HashMap<>();
+
+    static {
+        GROOVY_CALLSITE_INTERCEPTORS.put("getProperty", new SystemGetPropertyInterceptor());
+        GROOVY_CALLSITE_INTERCEPTORS.put("setProperty", new SystemSetPropertyInterceptor());
+
+        SystemGetPropertiesInterceptor getPropertiesInterceptor = new SystemGetPropertiesInterceptor();
+        GROOVY_CALLSITE_INTERCEPTORS.put("getProperties", getPropertiesInterceptor);
```

Comment:
💄 Perhaps move the list of applicable method names to each interceptor? It seems they are already tailored to specific methods anyway.

Reply:
I've tried to do that, and ended up with a quite involved solution. However, I fixed some dispatch-related corner cases along the way and added the ability to intercept constructors of several classes, so the additional complexity might be justified. PTAL.